src/utils.js

```diff
@@ -99,54 +107,68 @@ export class Coordinate {
    * @param {Action} action
    * @returns {Coordinate}
    */
-  translateByAction(action) {
+  translatedByAction(action) {
     const actionDelta = actionDeltas[action];
 
     if (action === undefined) {
       throw new TypeError(`The action "${action}" is invalid`);
     }
 
-    return this.translateByDelta(actionDelta);
+    return this.translatedByDelta(actionDelta);
   }
 
   /**
-   * Convert this Coordinate to an integer position
-   * @param {number} mapWidth
-   * @param {number} mapHeight
+   * Convert this Coordinate to an integer position in the map's coordinate system
+   * @param {PaintbotMap} map
    * @returns {number}
    */
-  toPosition(mapWidth, mapHeight) {
-    if (this.isOutOfBounds(mapWidth, mapHeight)) {
+  toPosition(map) {
+    if (this.isOutOfBounds(map)) {
       throw new RangeError('The coordinate must be within the bounds in order to convert to position');
     }
 
-    const { x, y } = this;
-    return x + y * mapWidth;
+    return this.x + this.y * map.width;
   }
 }
 
 export class MapUtility {
+  /** @type {PaintbotMap} */
+  #map;
+  /** @type {string} */
+  #playerId;
+  /** @type Map<string, CharacterInfo> */
+  #characters = new Map();
+  /** @private @type Map<number, Tile> */
+  #tiles = new Map();
+
+  get map() {
+    return this.#map;
+  }
+
   /**
    * @param {PaintbotMap} map
    * @param {string} playerId
    */
   constructor(map, playerId) {
-    this.map = map;
-    this.mapSize = map.height * map.width;
-    this.playerId = playerId;
-    this.characterInfoMap = new Map();
-    this.positionToTiles = new Map();
-    for (const p of this.map.powerUpPositions) {
-      this.positionToTiles.set(p, powerUpTile);
+    this.#map = map;
+    this.#playerId = playerId;
+
+    for (const powerUpPosition of map.powerUpPositions) {
+      this.#tiles.set(powerUpPosition, powerUpTile);
     }
 
-    for (const p of map.obstaclePositions) {
-      this.positionToTiles.set(p, obstactleTile);
+    for (const obstaclePosition of map.obstaclePositions) {
+      this.#tiles.set(obstaclePosition, obstactleTile);
     }
 
-    for (const ci of map.characterInfos) {
-      this.characterInfoMap.set(ci.id, ci);
-      this.positionToTiles.set(ci.position, createCharacterTile(ci));
+    for (const characterInfo of map.characterInfos) {
+      this.#characters.set(characterInfo.id, characterInfo);
+      this.#tiles.set(characterInfo.position, createCharacterTile(characterInfo.id));
+
+      const colourTile = createColourTile(characterInfo.id);
+      for (const colouredPosition of characterInfo.colouredPositions) {
+        this.#tiles.set(colouredPosition, colourTile);
```

Comment:
You realize that a bot can be standing on a coloured tile, I.e. the position blue is standing on will be coloured blue, so no you're overwritting the character tile with the colour tile